### PR TITLE
Implement algorithms task #5 - ENikonov [new unit tests]

### PR DESCRIPTION
**Pull request by re-nikonov**

### CODE REVIEW

File: algorithms task (5)/student/include/impl.h [hunk 1]
```diff
@@ -10,4 +13,12 @@
  * @return gathered range
  */
 template <typename BiIt, typename UnaryPredicate>
-std::pair<BiIt, BiIt> gather(BiIt begin, BiIt end, BiIt position, UnaryPredicate predicate);
+std::pair<BiIt, BiIt> gather(BiIt begin, BiIt end, BiIt position, UnaryPredicate predicate)
+{
+    const auto partitioningEnd = std::partition(begin, end, predicate);
+    const auto gatheringOffset = std::min(std::distance(begin, position),
```
Comment (smay1613) [inline algorithms task (5)/student/include/impl.h:19]: distance call + rotate is not effective operation here, please consider  simpler algorithm

Comment (re-nikonov) [reply algorithms task (5)/student/include/impl.h]: Found and implemented new algorithm.
Please take in account my new unit tests condition about that "right" element must be at specified position.

File: algorithms task (5)/student/include/impl.h [hunk 2]
```diff
@@ -10,4 +13,19 @@
...
+    {
+        return !predicate(value);
+    };
+
+    auto partitionBegin = std::partition(begin, std::next(position), notPredicate);
+    auto partitionEnd = std::partition(position, end, predicate);
+
+    if (partitionBegin == std::next(position))
+    {
+        std::advance(partitionBegin, -1);
```
Comment (smay1613) [inline algorithms task (5)/student/include/impl.h:28]: What if partition begin is at "begin"?

File: algorithms task (5)/student/include/impl.h [hunk 3]
```diff
@@ -10,4 +13,19 @@
...
+{
+    const auto notPredicate = [predicate](const typename std::iterator_traits<BiIt>::reference value)
+    {
+        return !predicate(value);
+    };
+
+    auto partitionBegin = std::partition(begin, std::next(position), notPredicate);
+    auto partitionEnd = std::partition(position, end, predicate);
+
+    if (partitionBegin == std::next(position))
```
Comment (smay1613) [inline algorithms task (5)/student/include/impl.h:26]: What if position is end?

File: algorithms task (5)/student/include/impl.h (hunk 2 above)
Comment (re-nikonov) [reply algorithms task (5)/student/include/impl.h:28]: Partition's begin cannot be "begin" here as it equals to std::next(position), so it will be at least "begin" + 1.

File: algorithms task (5)/student/include/impl.h (hunk 3 above)
Comment (re-nikonov) [reply algorithms task (5)/student/include/impl.h:26]: Undefined behavior, but I assumed that it doesn't make sense to pass end() as position. I think it is not so important here. Do you think it should be fixed?

Comment (smay1613) [reply algorithms task (5)/student/include/impl.h:26]: consider the empty range

File: algorithms task (5)/student/include/impl.h (hunk 2 above)
Comment (smay1613) [reply algorithms task (5)/student/include/impl.h:28]: please add tests for empty range

Comment (re-nikonov) [reply algorithms task (5)/student/include/impl.h]: Right! Nothing is added, the test for empty range already exists. But the algorithm is changed.

File: algorithms task (5)/student/include/impl.h (hunk 3 above)
Comment (re-nikonov) [reply algorithms task (5)/student/include/impl.h]: Fixed.

**Comment by re-nikonov** [comment]
Is this task accepted? Shall we consider this task as finished?

**Comment by smay1613** [comment]
Everything is fine here now
